core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2022,29 +2022,80 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    def arbiter(
+      promise: Promise[Cause[E], List[B]],
+      buffer: Ref[Array[B]],
+      todo: Ref[Int],
+      idx: Int
+    )(res: Exit[E, B]): ZIO[R, Nothing, Unit] =
+      res.foldM[R, Nothing, Unit](
+        e => promise.fail(e).unit,
+        a =>
+          todo.modify { t =>
+            val updateBuffer = buffer.modify(b => (b.update(idx, a), b))
```

Comment:
Using a mutable collection in `Ref` is pretty odd, I'm not sure there's any difference whatsoever here wrt a plain `val`, even volatility in Ref will apply to the array reference, not the contents of the array – since the reference does not change, it has no effect. Even `modify` here is completely equivalent to `buffer.get.map(b => b(idx) = a)` because `compareAndSet` spin-lock will never fail and retry, as `b` is always the same reference.

Reply:
I had same concern: https://github.com/zio/zio/pull/1773#discussion_r332588328 in original PR review. I think I'll revert this to `Array` or `mutable.ArraySeq` - depends on how I would handle error.